Mathlib/NumberTheory/ModularForms/EisensteinSeries/UniformConvergence.lean

```diff
@@ -0,0 +1,291 @@
+  fin_cases x <;> aesop
+
+/-This proof is faster than the one above, but longer. -/
+lemma ne_zero_if_max' {x : Fin 2 → ℤ} (hx : x ≠ 0)
+    (h : (max (x 0).natAbs (x 1).natAbs) = (x 1).natAbs) : (x 1) ≠ 0 := by
+  apply ne_zero_if_max (x := ![x 1, x 0]) ?_ (by simpa only [Fin.isValue, Matrix.cons_val_zero,
+    Matrix.cons_val_one, Matrix.head_cons, max_eq_left_iff, max_eq_right_iff] using h)
+  simp only [ne_eq, Matrix.cons_eq_zero_iff, Matrix.zero_empty, and_true, not_and]
+  intro h1 h0
+  rw [Function.ne_iff, Fin.exists_fin_two, h1, h0] at hx
+  simp only [Pi.zero_apply, ne_eq, not_true_eq_false, or_self] at *
+
+lemma div_max_sq_ge_one (x : Fin 2 → ℤ) (hx : x ≠ 0) :
+    (1 : ℝ) ≤ (x 0 / (max (x 0).natAbs (x 1).natAbs)) ^ 2 ∨
+      (1 : ℝ) ≤ (x 1 / (max (x 0).natAbs (x 1).natAbs)) ^ 2 := by
+  refine (max_choice (x 0).natAbs (x 1).natAbs).imp (fun H1 => ?_) (fun H2 => ?_)
+  · simp only [Fin.isValue, H1, Int.cast_natAbs, Int.cast_abs, div_pow, _root_.sq_abs, ne_eq,
+    OfNat.ofNat_ne_zero, not_false_eq_true, pow_eq_zero_iff, Int.cast_eq_zero, ne_zero_if_max hx H1,
+    div_self, le_refl]
+  · simp only [Fin.isValue, H2, Int.cast_natAbs, Int.cast_abs, div_pow, _root_.sq_abs, ne_eq,
+    OfNat.ofNat_ne_zero, not_false_eq_true, pow_eq_zero_iff, Int.cast_eq_zero,
+    ne_zero_if_max' hx H2, div_self, le_refl]
+
+lemma rpow_bound {k : ℝ} (hk : 0 ≤ k) (z : ℍ) (x : Fin 2 → ℤ) (hx : x ≠ 0) :
+    ((r z) ^ k) * (max (x 0).natAbs (x 1).natAbs) ^ k ≤
+      (Complex.abs ((x 0 : ℂ) * (z : ℂ) + (x 1 : ℂ))) ^ k := by
+  by_cases hk0 : k ≠ 0
+  · let n := max (x 0).natAbs (x 1).natAbs
+    have hn0 : n ≠ 0 := by
+      rw [← Iff.ne ((eq_zero_iff_eq_zero_of_mem_box (α := ℤ × ℤ) (x := ((x 0, x 1)))) (by simp)),
+        ← Iff.ne (Function.Injective.eq_iff (Equiv.injective (piFinTwoEquiv fun _ ↦ ℤ)))] at *
+      simpa only [Fin.isValue, ne_eq, Prod.mk_eq_zero, not_and, piFinTwoEquiv_apply, Pi.zero_apply,
+        Prod.mk.injEq] using hx
+    have h11 : ((x 0) * ↑z + (x 1)) =
+        (((x 0 : ℝ) / (n : ℝ)) * (z : ℂ) + (x 1 : ℝ) / (n : ℝ)) * ((n : ℝ)) := by
+        have : (n : ℂ) ≠ 0 := by norm_cast
+        field_simp [this]
+    simp only [Nat.cast_max, h11, map_mul, abs_ofReal, ge_iff_le]
+    rw [Real.mul_rpow (by apply apply_nonneg) (by apply abs_nonneg)]
+    cases' (div_max_sq_ge_one x hx) with H1 H2
+    · apply mul_le_mul _ (by norm_cast) _ (by apply Real.rpow_nonneg (Complex.abs.nonneg _) k)
+      · simpa only [Fin.isValue, ofReal_intCast, Nat.cast_max, ofReal_div, n] using
+        (Real.rpow_le_rpow (r_pos _).le (auxbound1 z (x 1 / n) H1) hk)
+      · positivity
+    · apply mul_le_mul _ (by norm_cast) _ (by apply Real.rpow_nonneg (Complex.abs.nonneg _) k)
+      · simpa only [Fin.isValue, ofReal_intCast, Nat.cast_max, ofReal_div, n] using
+        (Real.rpow_le_rpow (r_pos _).le (auxbound2 z (x 0 / n) H2) hk)
+      · positivity
+  · simp only [ne_eq, not_not] at hk0
+    simp only [hk0, Real.rpow_zero, Nat.cast_max, mul_one, le_refl]
```

Comment:
- There's no need to handle the case `k = 0` separately; the general-case proof works fine for `k = 0` also.
- Please don't use type annotations like `(x 0 : ℂ)` unnecessarily. Lean is _very_ good at guessing types for itself, and you'll hurt its feelings if you treat it as if it's stupid. :-). I have cultivated a habit that before PR'ing any code, I go through every type annotation and try deleting it to see if the code still works.
- You have an absolute epidemic of redundant brackets (here and in many other places in the code).
- You seem to be jumping through hoops in order to use your lemma `eq_zero_iff_eq_zero_of_mem_box`, in a situation where it's easier just to prove the result directly.
- The `field_simp` call takes about 0.7 sec on my machine, which is an eternity given how fast modern computers are! The overhead of firing up `field_simp` is so great that it's absolutely not worth using it for simple little goals like this one which can be `rw`'ed in a few steps by hand.
- I removed many type annotations from `h11`. The new expression elaborates slightly differently (replacing a two-step cast  `ℕ → ℝ → ℂ` with `ℕ → ℂ`), but the difference is a gain, not a loss, as it simplifies arguments further down.
```suggestion
lemma rpow_bound {k : ℝ} (hk : 0 ≤ k) (z : ℍ) (x : Fin 2 → ℤ) (hx : x ≠ 0) :
    (r z) ^ k * (max (x 0).natAbs (x 1).natAbs) ^ k ≤ Complex.abs (x 0 * z + x 1) ^ k := by
  rw [← Real.mul_rpow (r_pos _).le (Nat.cast_nonneg _)]
  refine Real.rpow_le_rpow (mul_nonneg (r_pos _).le (Nat.cast_nonneg _)) ?_ hk
  let n := max (x 0).natAbs (x 1).natAbs
  have hn0 : n ≠ 0 := by
    contrapose! hx
    rw [Nat.max_eq_zero_iff, Int.natAbs_eq_zero, Int.natAbs_eq_zero] at hx
    exact funext fun j ↦ by fin_cases j <;> tauto
  have h11 : x 0 * (z : ℂ) + x 1 = (x 0 / n * z + x 1 / n) * n := by
    push_cast
    rw [div_mul_eq_mul_div, ← add_div, div_mul_cancel₀ _ (Nat.cast_ne_zero.mpr hn0)]
  rw [Nat.cast_max, h11, map_mul, abs_natCast]
  gcongr
  · rcases div_max_sq_ge_one x hx with H1 | H2
    · simpa only [Fin.isValue, ofReal_div, ofReal_intCast, n] using auxbound1 z (x 1 / n) H1
    · simpa only [Fin.isValue, ofReal_div, ofReal_intCast, n] using auxbound2 z (x 0 / n) H2
  · norm_cast
```